lib/kdl/include/kdl/compact_trie.h

```diff
@@ -0,0 +1,714 @@
+             */
+            bool set_matched(const node* n) {
+                auto it = m_state.find(n);
+                assert(it != std::end(m_state));
+
+                auto& state = it->second;
+                if (state.node_matched) {
+                    return false;
+                }
+
+                state.node_matched = true;
+                if (state.fully_matched_children == n->m_children.size()) {
+                    // update the subtree match counts of all nodes on the path to the given node
+                    update_parent_states(state.parent);
+                }
+
+                return true;
+            }
+        private:
+            void update_parent_states(const node* n) {
+                while (n != nullptr) {
+                    auto it = m_state.find(n);
+                    assert(it != std::end(m_state));
+
+                    auto& state = it->second;
+                    state.fully_matched_children += 1u;
+                    if (!state.node_matched || state.fully_matched_children < n->m_children.size()) {
+                        // parent is not fully matched, so it cannot contribute to its parents' subtree match count yet
+                        break;
+                    }
+
+                    n = it->second.parent;
+                }
+            }
+        };
+
+        /**
+         * A trie node. Children are stored in a set ordered by `node_cmp`. Each node can store a given value multiple
+         * times.
+         *
+         * A trie node has only const methods and therefore appears to be immutable, but this is not the case. All of
+         * its members are actually declared as mutable and may be modified by the `insert` or `remove` member
+         * functions. The reason for this peculiar design choice is that trie nodes are stored by value in a `std::set`,
+         * and `std::set` does not allow its contents to be modified, since that might affect the position of the value
+         * in the set.
+         *
+         * But in this particular case, the position of a node in the containing set depends only on the node's key, and
+         * the node's key is changed only in ways that does not affect its position. The key can only become shorter
+         * or longer, that is, key "abc" might be changed to "ab" or "abcd", and neither of these changes affects how
+         * the key compares to other keys due to the order introduce by `node_cmp`, which consideres two keys equivalent
+         * if they share a non-empty prefix.
+         *
+         * This implies that the following two conditions hold for a node with key "abc":
+         * - There is no sibling node with a key that is a prefix of "abc".
+         * - There is no sibling node with a key that that has "abc" as a prefix.
+         *
+         * Therefore, shortening or extending a node's key does not affect its position in the containing set, and we
```

Comment:
This documentation is pretty confusing - if my reading of node_cmp is correct, what's actually going on is the first char of the node's text is used as the key for the std::set. Because of this, the characters after the first one are allowed to change, as long as the first one doesn't.

Reply:
Good observation! I have changed the implementation accordingly.